File: .NET/VS2010TrainingKit/Labs/WhatsNewInWCF4/Source/Ex7-DiscoveryProxy/End/VB/ChatProxy/FindAsyncResult.vb
' ----------------------------------------------------------------------------------
' Microsoft Developer & Platform Evangelism
' 
' Copyright (c) Microsoft Corporation. All rights reserved.
' 
' THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
' OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
' ----------------------------------------------------------------------------------
' The example companies, organizations, products, domain names,
' e-mail addresses, logos, people, places, and events depicted
' herein are fictitious.  No association with any real company,
' organization, product, domain name, email address, logo, person,
' places, or events is intended or should be inferred.
' ----------------------------------------------------------------------------------

Imports System.Collections.ObjectModel
Imports System.ServiceModel.Discovery

Public Class FindAsyncResult
    Inherits CompletedAsyncResult
    Private matchingEndpoints As Collection(Of EndpointDiscoveryMetadata)

    Public Sub New(ByVal matchingEndpoints As Collection(Of EndpointDiscoveryMetadata), ByVal callback As AsyncCallback, ByVal state As Object)
        MyBase.New(callback, state)
        Me.matchingEndpoints = matchingEndpoints
    End Sub

    ' Hides the inherited End from CompletedAsyncResult
    ' This method returns a collection of metadata
    Public Shared Shadows Function [End](ByVal result As IAsyncResult) As Collection(Of EndpointDiscoveryMetadata)
        Dim thisPtr = AsyncResult.End(Of FindAsyncResult)(result)
        Return thisPtr.matchingEndpoints
    End Function
End Class

File: .NET/VS2010TrainingKit/Labs/WhatsNewInWCF4/Source/Assets/ChatProxy/C#/FindAsyncResult.cs
// ----------------------------------------------------------------------------------
// Microsoft Developer & Platform Evangelism
// 
// Copyright (c) Microsoft Corporation. All rights reserved.
// 
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
// OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
// ----------------------------------------------------------------------------------
// The example companies, organizations, products, domain names,
// e-mail addresses, logos, people, places, and events depicted
// herein are fictitious.  No association with any real company,
// organization, product, domain name, email address, logo, person,
// places, or events is intended or should be inferred.
// ----------------------------------------------------------------------------------

namespace ChatProxy
{
    using System;
    using System.Collections.ObjectModel;
    using System.ServiceModel.Discovery;
    
    public class FindAsyncResult : CompletedAsyncResult
    {
        private Collection<EndpointDiscoveryMetadata> matchingEndpoints;

        public FindAsyncResult(
            Collection<EndpointDiscoveryMetadata> matchingEndpoints,
            AsyncCallback callback,
            object state)
            : base(callback, state)
        {
            this.matchingEndpoints = matchingEndpoints;
        }

        // Hides the inherited End from CompletedAsyncResult
        // This method returns a collection of metadata
        public static new Collection<EndpointDiscoveryMetadata>
              End(IAsyncResult result)
        {
            FindAsyncResult thisPtr = AsyncResult.End<FindAsyncResult>(result);
            return thisPtr.matchingEndpoints;
        }
    }
}